driver-core/src/main/com/mongodb/internal/connection/BaseCluster.java

```diff
@@ -105,34 +121,42 @@ public ServerTuple selectServer(final ServerSelector serverSelector, final Opera
         try {
             CountDownLatch currentPhase = phase.get();
             ClusterDescription curDescription = description;
+            logServerSelectionStarted(operationContext, serverSelector, curDescription);
             ServerSelector compositeServerSelector = getCompositeServerSelector(serverSelector);
             ServerTuple serverTuple = selectServer(compositeServerSelector, curDescription);
 
-            boolean selectionFailureLogged = false;
+            boolean selectionWaitingLogged = false;
 
             long startTimeNanos = System.nanoTime();
             long curTimeNanos = startTimeNanos;
-            long maxWaitTimeNanos = getMaxWaitTimeNanos();
+            Long maxWaitTimeNanos = getMaxWaitTimeNanos();
```

Comment:
Does this method need to be refactored in order to accomplish the logging spec changes? Note that much of this has been refactored already in the CSOT branch. It seems that we only need to either add methods that log messages, or remove existing logs.

Reply:
- `remainingTimeNanos` has to be computed and logged, unless it's infinite. This required changing the code.
- Creating exceptions and logging them is related, which means that the line `throwIfIncompatible(curDescription);` needed to be changed.

Reply:
> Note that much of this has been refactored already in the CSOT branch.

I won't be surprised if merging this code from `master` and CSOT will effectively require reintroducing the spec changes in the `selectServer` methods (not from scratch, though, as the methods that create exceptions and the methods that log them don't need changes)